### PR TITLE
Elected summarizer client occasionally does not start summarizer.

### PR DESCRIPTION
**Pull request by vladsud**
Issue #7311: Summarizer election may result in summarizer not being started 
Much more details are in https://onedrive.visualstudio.com/SPIN/_queries/edit/1186576/?triage=true

The core of the problem that client who is assigned to be a summarizer never creates summarizing client.
So I've been doing a bunch of code reviewing to figure out all possible reasons for why it may happen.
But looking back at history for this container, it actually was elected summarizer before, and summarizer started, but never finished and there were no events regarding the end. So that pointed to it never finishing booting in the first place.
So from code, it looks like a condition that was never handled correctly - container being disposed before summarizer even had a chance to start. So addressing this.

Also while reviewing all the ways how we propagate events RE leader election, we might run into state where events and state is not synchronized due to an early return statement in summarizerClientElection.ts. Fixing that too.

### CODE REVIEW

**Comment by fluidgit** [comment]
<details><summary><b><span style="color: green">■</span> @fluidframework/base-host:</b> No change</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr>
    <td>main.js</td>
    <td>185.72 KB</td>
    <td>185.72 KB</td>
    <td><span style="color: green">■</span> No change</td>
  </tr><tr>
    <td>Total Size</td>
    <td>213.31 KB</td>
    <td>213.31 KB</td>
    <td><span style="color: green">■</span> No change</td>
  </tr></table></details><details><summary><b><span style="color: coral">⯅</span> @fluid-example/bundle-size-tests:</b> +193 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr>
    <td>container.js</td>
    <td>163.01 KB</td>
    <td>163.2 KB</td>
    <td><span style="color: coral">⯅</span> +193 Bytes</td>
  </tr><tr>
    <td>map.js</td>
    <td>44.71 KB</td>
    <td>44.71 KB</td>
    <td><span style="color: green">■</span> No change</td>
  </tr><tr>
    <td>matrix.js</td>
    <td>143.51 KB</td>
    <td>143.51 KB</td>
    <td><span style="color: green">■</span> No change</td>
  </tr><tr>
    <td>odspDriver.js</td>
    <td>173.42 KB</td>
    <td>173.42 KB</td>
    <td><span style="color: green">■</span> No change</td>
  </tr><tr>
    <td>odspPrefetchSnapshot.js</td>
    <td>38.7 KB</td>
    <td>38.7 KB</td>
    <td><span style="color: green">■</span> No change</td>
  </tr><tr>
    <td>sharedString.js</td>
    <td>165.09 KB</td>
    <td>165.09 KB</td>
    <td><span style="color: green">■</span> No change</td>
  </tr><tr>
    <td>Total Size</td>
    <td>761.13 KB</td>
    <td>761.31 KB</td>
    <td><span style="color: coral">⯅</span> +193 Bytes</td>
  </tr></table></details><hr><p>Baseline commit: 406d2fb0ec90580b6e1c3900cebc7b3211b13f7a</p>
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 3e98ec6c385b720aeb667702878e5242da9161d9
</p>

File: packages/runtime/container-runtime/src/runWhileConnectedCoordinator.ts [hunk 1]
```diff
@@ -68,9 +68,9 @@ export class RunWhileConnectedCoordinator implements ICancellableSummarizerContr
         if (this.runtime.connected) {
             this.everConnected = true;
         } else if (this.runtime.disposed) {
+            this.everConnected = true;
```
Comment (andre4i) [inline packages/runtime/container-runtime/src/runWhileConnectedCoordinator.ts:71]: Not saying that this is wrong but I'm a bit confused about why this is a safe inference? Would this whole thing be moot if the container runtime was never connected?

Comment (vladsud) [reply packages/runtime/container-runtime/src/runWhileConnectedCoordinator.ts:71]: I like to have fewer different states to reason. Given that disposed and previously connected, and disposed and never connected should be treated the same by this code, I prefer to collapse these two states in terms of how they are represented in states of this class.

That said, I'm realizing that it's not actually true and current logic still has pretty big gap - it never listens on "disposed" state. :) So changes are required.

Comment (vladsud) [reply packages/runtime/container-runtime/src/runWhileConnectedCoordinator.ts:71]: And opened https://github.com/microsoft/FluidFramework/issues/7338 to follow up with proper UTs here, as obviously it was a huge gap in this logic.

File: packages/runtime/container-runtime/src/runWhileConnectedCoordinator.ts [hunk 2]
```diff
@@ -88,14 +88,16 @@ export class RunWhileConnectedCoordinator implements ICancellableSummarizerContr
...
     }
 
     /**
      * Starts and waits for a promise which resolves when connected.
      * The promise will also resolve if stopped either externally or by disconnect.
      */
-    public async waitStart() {
+    protected async waitStart() {
         if (!this.runtime.connected && !this.everConnected) {
+            assert(!this._cancelled, "not cancelled");
```
Comment (pleath) [inline packages/runtime/container-runtime/src/runWhileConnectedCoordinator.ts:100]: Nit: could the assert message be more explanatory?

**Review by pleath — APPROVED** [review]

File: packages/runtime/container-runtime/src/runWhileConnectedCoordinator.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/runWhileConnectedCoordinator.ts:100]: Actually, on second thought, I do not think this assert adds anything (it just asserts current structure of the code, but it if changes, this method would still be correct). So I'll remove it.

**Comment by vladsud** [comment]
@pleath, @andre4i - thanks for feedback, this was useful, forced me to refactor code a bit more to make it both easier to follow but also have less chances to have bugs (now and later). Please glance one more time - latest changes are only in runWhileConnectedCoordinator.ts. I moved all the code from constructor into waitStarted() method and that allowed me to make it more streamlined. Running stress tests with it - looks good so far :)

**Review by andre4i — APPROVED** [review]

**Pull request merged**